src/main/java/br/ufpe/cin/mergers/SemistructuredMerge.java

```diff
@@ -43,30 +53,37 @@
 	 * @throws TextualMergeException
 	 */
 	public static String merge(File left, File base, File right, MergeContext context)	throws SemistructuredMergeException, TextualMergeException {
-		try {
-			// parsing the files to be merged
-			JParser parser = new JParser();
-			FSTNode leftTree = parser.parse(left);
-			FSTNode baseTree = parser.parse(base);
-			FSTNode rightTree = parser.parse(right);
-
-			// merging
-			context.join(merge(leftTree, baseTree, rightTree));
-
-			// handling special kinds of conflicts
-			ConflictsHandler.handle(context);
-
-		} catch (ParseException | FileNotFoundException | UnsupportedEncodingException | TokenMgrError ex) {
-			String message = ExceptionUtils.getCauseMessage(ex);
-			if(ex instanceof FileNotFoundException) //FileNotFoundException does not support custom messages
-				message = "The merged file was deleted in one version.";
-			throw new SemistructuredMergeException(message, context);
-		}
-
-		// during the parsing process, code indentation is typically lost, so we reindent the code
-		return FilesManager.indentCode(Prettyprinter.print(context.superImposedTree));
+		return merge(left, base, right, context, CONFLICT_HANDLERS);
 	}
 
+    public static String merge(File left, File base, File right, MergeContext context, List<ConflictHandler> conflictHandlers)	throws SemistructuredMergeException, TextualMergeException {
+        try {
+            // parsing the files to be merged
+            JParser parser = new JParser();
+            FSTNode leftTree = parser.parse(left);
+            FSTNode baseTree = parser.parse(base);
+            FSTNode rightTree = parser.parse(right);
+
+            // merging
+            context.join(merge(leftTree, baseTree, rightTree));
+
+            // handling special kinds of conflicts
+            context.semistructuredOutput = Prettyprinter.print(context.superImposedTree); //partial result of semistructured merge is necessary for further processing
+            for (ConflictHandler conflictHandler : conflictHandlers) {
+                conflictHandler.handle(context);
```

Comment:
failure of one handler implies failures of the others? is that the intention?

Reply:
I never thought about that, but they actually don't fail (they don't have exceptions). So far the only possible failure is that from TextualMerge, and in that case, yes, because the merge is broken.
The handlers have implicit pre-conditions, if these conditions are not met, they simple do nothing.

Reply:
we better double check if the handlers have no chances of raising exceptions (runtime, for example)

Reply:
What should happen if a handler throws a runtime exception, then?

Reply:
It should be caught and chained by SemistructuredMerge as SemistructuredMergeException in https://github.com/guilhermejccavalcanti/jFSTMerge/blob/master/src/main/java/br/ufpe/cin/mergers/SemistructuredMerge.java#L59 but isn't doing that I guess. Luck me that so far the handlers have never failed :-)

Reply:
@guilhermejccavalcanti forgetting about the existing handlers, focusing only on the abstraction. suppose we have 3 handlers. if the second fails, raising an exception, should the third one be executed? that's what @gaabs needs to know

Reply:
Currently, if one fails, the remaining won't be executed. Conceptually, I think they are independent. One should not influence the other. I think the these handllers are good example of the Decorator Pattern (https://en.wikipedia.org/wiki/Decorator_pattern , although I have not implemented then with that in mind)

Reply:
I also think they should be independent, but that's not what is reflected in the implementation. OK @gaabs?